boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -256,19 +260,35 @@ UefiMemGetMemoryMap(ULONG *MemoryMapSize)
             }
         }
 
-        UefiSetMemory(FreeldrMem,
-                      MapEntry->PhysicalStart,
-                      MapEntry->NumberOfPages,
-                      MemoryType);
+        /* Sometimes our loader can be loaded into higher memory than we ever allocate */
+        if (MemoryType == LoaderLoadedProgram)
+        {
+            if (((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT) > LoaderPagesSpanned)
+            {
+                /* This value needs to be adjusted if this occurs */
+                LoaderPagesSpanned = ((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT);
```

Comment:
The comments don't really explain what this hack is trying to achieve (counteracting what? etc.). You've just told _**in which case**_ it is needed.

Reply:
How is it a hack? The loader spans the pages of the LoaderLoadedProgram? This is exactly what BIOS does but it knows it’s full range immediately

Reply:
Ah yes thanks for pointing. Actually this variable is indeed checked (and re-calculated if needed) like every time there is a memory allocation.